src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1195,17 +1463,735 @@ private static void ReportReadOnlyError(Symbol symbol, SyntaxNode node, BindValu
 
             ErrorCode[] ReadOnlyErrors =
             {
-                ErrorCode.ERR_RefReadonlyNotField,
                 ErrorCode.ERR_RefReturnReadonlyNotField,
+                ErrorCode.ERR_RefReadonlyNotField,
                 ErrorCode.ERR_AssignReadonlyNotField,
-                ErrorCode.ERR_RefReadonlyNotField2,
                 ErrorCode.ERR_RefReturnReadonlyNotField2,
+                ErrorCode.ERR_RefReadonlyNotField2,
                 ErrorCode.ERR_AssignReadonlyNotField2,
             };
 
-            int index = (checkingReceiver ? 3 : 0) + (RequiresRefOrOut(kind) ? 0 : (kind == BindValueKind.RefReturn ? 1 : 2));
+            int index = (checkingReceiver ? 3 : 0) + (kind == BindValueKind.RefReturn ? 0 : (RequiresRefOrOut(kind) ? 1 : 2));
             Error(diagnostics, ReadOnlyErrors[index], node, symbolKind, symbol);
         }
 
+        /// <summary>
+        /// Checks whether given expression can escape from the current scope to the `escapeTo`
+        /// In a case if it cannot a bad expression is returned and diagnostics is produced.
+        /// </summary>
+        internal BoundExpression ValidateEscape(BoundExpression expr, uint escapeTo, bool isByRef, DiagnosticBag diagnostics)
```

Comment:
Nit: rename ValidateEscape to `CheckEscape` to align with CheckRefEscape, CheckValEscape and other "Check" methods?

Reply:
ValidateEscape  is the entry point into this and is different as it will not only produce diagnostics, but also will return BadExpression in a case of a failure. I wanted it to not be mixed with CheckXXX methods. "Validate" is the best I could come up with. It may not be the best, but it is also intentionally different from "Check".
The other one I thought about was "Ensure", but it felt not much better.

---
In reply to: [135401525](https://github.com/dotnet/roslyn/pull/21334#discussion_r135401525) [](ancestors = 135401525)

Reply:
I'll live it ValidateEscape. There is a comment on what it does and I can't come up with a better name and it shoudl not be "Check"

---
In reply to: [135646445](https://github.com/dotnet/roslyn/pull/21334#discussion_r135646445) [](ancestors = 135646445,135401525)